### PR TITLE
os: uptime

### PR DESCRIPTION
**Pull request by indutny**
Cross-platform implementation and test.

### CODE REVIEW

**Comment by indutny** [comment]
Just tested on windows 7 - works fine and reports data close to GetTickCount().

File: src/unix/cygwin.c [hunk 1]
```diff
@@ -53,6 +54,22 @@ int uv_exepath(char* buffer, size_t* size) {
...
+uint64_t uv_uptime(void) {
+  uint64_t amount;
+  char line[512];
+  FILE *fpUptime = fopen("/proc/uptime", "r");
+
+  amount = 0;
+
+  if (fpUptime) {
+    if (fgets(line, 511, fpUptime) != NULL) {
+      sscanf(line, "%d.%d %d.%d", &amount);
```
Comment (bnoordhuis) [inline src/unix/cygwin.c]: Memory corruption.

File: src/unix/cygwin.c [hunk 2]
```diff
@@ -65,4 +82,4 @@ int uv_fs_event_init(uv_loop_t* loop,
 
 void uv__fs_event_destroy(uv_fs_event_t* handle) {
   assert(0 && "implement me");
-}
+*}
```
Comment (bnoordhuis) [inline src/unix/cygwin.c]: What's this?

File: src/unix/linux.c [hunk 3]
```diff
@@ -64,6 +69,23 @@ int uv_exepath(char* buffer, size_t* size) {
   return 0;
 }
 
+uint64_t uv_uptime(void) {
+#ifdef UV_USE_MONOTONIC
+  struct timespec now;
+  if (0 == clock_gettime(CLOCK_MONOTONIC, &now)) {
+    double uptime = now.tv_sec;
+    uptime += (double)now.tv_nsec / 1000000000.0;
```
Comment (bnoordhuis) [inline src/unix/linux.c]: Don't use floating point here.

File: src/unix/sunos.c [hunk 4]
```diff
@@ -62,6 +64,30 @@ int uv_exepath(char* buffer, size_t* size) {
...
+uint64_t uv_uptime(void) {
+  kstat_ctl_t   *kc;
+  kstat_t       *ksp;
+  kstat_named_t *knp;
+
+  long hz = sysconf(_SC_CLK_TCK);
+  ulong_t clk_intr;
+
+  if ((kc = kstat_open()) == NULL)
+    throw "could not open kstat";
```
Comment (bnoordhuis) [inline src/unix/sunos.c]: Don't throw. This is C, not C++.

File: src/win/util.c [hunk 5]
```diff
@@ -94,3 +95,123 @@ done:
...
+  DWORD size = 4096;
+  DWORD readSize = 0;
+  LONG status = ERROR_MORE_DATA;
+  LPBYTE data = NULL;
+
+  do {
+    if (readSize > size) {
+      size = readSize;
+    } else {
+      readSize = size += 4096;
```
Comment (bnoordhuis) [inline src/win/util.c]: Break this up in two statements.

File: src/win/util.c [hunk 6]
```diff
@@ -94,3 +95,123 @@ done:
...
+
+    if (data == NULL) {
+      data = malloc(size);
+      // Handle malloc's fails
+      if (data == NULL) return NULL;
+    } else {
+      LPBYTE newData = realloc(data, size);
+      if (newData == NULL) break;
+      data = newData;
+    }
```
Comment (bnoordhuis) [inline src/win/util.c]: Don't need to branch on `data == NULL`

`realloc(NULL, size)` is equivalent to `malloc(size)`

File: src/win/util.c [hunk 7]
```diff
@@ -94,3 +95,123 @@ done:
...
+  } else {
+    free(data);
+    return NULL;
+  }
+}
+
+DWORD uv__lookup_counter_title_index(char* table_, char* title) {
+  char* table = table_;
+
+  // Skip first record and it's length
```
Comment (bnoordhuis) [inline src/win/util.c]: Don't use C99 / C++ style comments.

File: src/win/util.c [hunk 8]
```diff
@@ -94,3 +95,123 @@ done:
...
+
+uint64_t uv_uptime(void) {
+  // Receive performance counters
+  PPERF_DATA_BLOCK header =
+      (PPERF_DATA_BLOCK) uv__query_reg_data(HKEY_PERFORMANCE_DATA, TEXT("2"));
+
+  if (header == NULL) return -1;
+
+  // Receive performance counter's titles by ids
+  char* table = (char*) uv__query_reg_data(HKEY_PERFORMANCE_DATA,
```
Comment (bnoordhuis) [inline src/win/util.c]: Variable declarations should go at the top.

File: src/win/util.c [hunk 9]
```diff
@@ -94,3 +95,123 @@ done:
...
+  }
+
+  // Find first object
+  PPERF_OBJECT_TYPE object = (PPERF_OBJECT_TYPE) ((LPBYTE) header +
+                                                  header->HeaderLength);
+
+  // If we don't have any objects or object has no counters - return error
+  if (header->NumObjectTypes == 0 || object->NumCounters == 0) {
+    free(header);
+    return -1;
```
Comment (bnoordhuis) [inline src/win/util.c]: I would like it better if this function had a single clean up and exit point (use `goto`), makes it easier to follow and check for resource leaks.

@piscisaureus: You should probably review it as well. It looks okay to me but I'm not a windows guru.

File: src/unix/cygwin.c (hunk 2 above)
Comment (indutny) [inline src/unix/cygwin.c]: Secret C technique :)

**Comment by indutny** [comment]
Fixed all errors and rebased on master

**Comment by indutny** [comment]
Tested on mingw and linux.

File: src/win/util.c (hunk 9 above)
Comment (piscisaureus) [inline src/win/util.c]: @bnoordhuis @indutny

I agree on @bnoordhuis' comment. Otherwise LGTM.

Comment (indutny) [inline src/win/util.c]: Please take a look at latest version, fixed here.

**Comment by indutny** [comment]
@bnoordhuis fixed.

**Comment by indutny** [comment]
@bnoordhuis please review, should be fixed now

**Comment by bnoordhuis** [comment]
Sorry Fedor, @igorzi already implemented this.

**Pull request closed without merging**